src/EnergyPlus/api/datatransfer.cc

```diff
@@ -573,6 +582,50 @@ Real64 getInternalVariableValue(EnergyPlusState state, int handle)
     return 0;
 }
 
+int getEMSGlobalVariableHandle(EnergyPlusState state, const char *name)
+{
+    auto *thisState = static_cast<EnergyPlus::EnergyPlusData *>(state);
+    int index = 0;
+    for (auto const &erlVar : thisState->dataRuntimeLang->ErlVariable) {
+        index++;
+        if (EnergyPlus::Util::SameString(name, erlVar.Name)) {
+            return index;
+        }
+    }
+    return 0;
+}
+
+Real64 getEMSGlobalVariableValue(EnergyPlusState state, int handle)
+{
+    auto *thisState = static_cast<EnergyPlus::EnergyPlusData *>(state);
+    if (handle < 0 || handle > thisState->dataRuntimeLang->NumErlVariables) {
+        // need to fatal out once the process is done
+        // throw an error, set the fatal flag, and then return 0
+        EnergyPlus::ShowSevereError(
+            *thisState, fmt::format("Data Exchange API: Problem -- index error in getEMSGlobalVariableValue; received handle: {}", handle));
+        EnergyPlus::ShowContinueError(
+            *thisState, "The getEMSGlobalVariableValue function will return 0 for now to allow the process to finish, then EnergyPlus will abort");
+        thisState->dataPluginManager->apiErrorFlag = true;
+        return 0;
+    }
+    return thisState->dataRuntimeLang->ErlVariable(handle).Value.Number;
+}
+
+void setEMSGlobalVariableValue(EnergyPlusState state, int handle, Real64 value)
+{
+    auto *thisState = static_cast<EnergyPlus::EnergyPlusData *>(state);
+    if (handle < 0 || handle > thisState->dataRuntimeLang->NumErlVariables) {
+        // need to fatal out once the plugin is done
+        // throw an error, set the fatal flag, and then return
+        EnergyPlus::ShowSevereError(
+            *thisState, fmt::format("Data Exchange API: Problem -- index error in setEMSGlobalVariableValue; received handle: {}", handle));
+        EnergyPlus::ShowContinueError(*thisState,
+                                      "The setEMSGlobalVariableValue function will return to allow the plugin to finish, then EnergyPlus will abort");
+        thisState->dataPluginManager->apiErrorFlag = true;
+    }
+    thisState->dataRuntimeLang->ErlVariable(handle).Value.Number = value;
+}
```

Comment:
Here, I'm pretty sure this is not ok.

ErlVariable array contains a bunch of builtins that you aren't supposed to be able to mess with. NullVariableNum, YEAR, FALSE, TRUE, PI, ACTUALTIME, etc.

Reply:
If you haven't defined any EMS stuff in your IDF, this is what the ErlVariable array contains:

```
1 - NULL
2 - FALSE
3 - TRUE
4 - OFF
5 - ON
6 - PI
7 - TIMESTEPSPERHOUR
8 - YEAR
9 - CALENDARYEAR
10 - MONTH
11 - DAYOFMONTH
12 - DAYOFWEEK
13 - DAYOFYEAR
14 - HOUR
15 - TIMESTEPNUM
16 - MINUTE
17 - HOLIDAY
18 - DAYLIGHTSAVINGS
19 - CURRENTTIME
20 - SUNISUP
21 - ISRAINING
22 - SYSTEMTIMESTEP
23 - ZONETIMESTEP
24 - CURRENTENVIRONMENT
25 - ACTUALDATEANDTIME
26 - ACTUALTIME
27 - WARMUPFLAG
```

Reply:
Most of these (except the constants such as TRUE, FALSE) are being assigned a value in InitializeRuntimeLanguage https://github.com/NREL/EnergyPlus/blob/52fee4b00a1d679d545dfaf69a3704227b11ff4c/src/EnergyPlus/RuntimeLanguageProcessor.cc#L183

So a change would be therefore limited to any calling points after the point where you are setting a new value. (eg change that in BeginNewEnvironement calling point).

Not sure whether that makes a ton of sense. If not, maybe considering slicing the ErlVariable array to ommit the built in ones, so the user ones only are left.

Reply:
first 27 values in ErlVariable() are constant and built-in variables

Reply:
EMS Sensors, Actuators and InternalVariables declared in the IDF are actually emplaced before the built-in ones.

---

GetEMSInput is called before https://github.com/NREL/EnergyPlus/blob/52fee4b00a1d679d545dfaf69a3704227b11ff4c/src/EnergyPlus/EMSManager.cc#L603

So Sensors, Actuators and InternalVariables are loaded in the array first, and only then the InitializeRuntimeLanguage is called
https://github.com/NREL/EnergyPlus/blob/52fee4b00a1d679d545dfaf69a3704227b11ff4c/src/EnergyPlus/EMSManager.cc#L839-L840